### PR TITLE
[Ready for review] Redis Client implementation

### PR DESCRIPTION
**Pull request by hovsepm**

### CODE REVIEW

**Comment by hovsepm** [comment]
PR fails due to Redis Backend bug in Header location in Put operation.

**Comment by jianghaolu** [comment]
There are checkstyle errors: https://travis-ci.org/Azure/azure-sdk-for-java/builds/159680208#L3234

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+
+    /**
+     * @return the enableNonSslPort value
+     */
+    Boolean enableNonSslPort();
+
+    /**
+     * @return the shardCount value
+     */
+    Integer shardCount();
```
Comment (anudeepsharma) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: Is shardCount typo?

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+
+    /**
+     * Regenerates the access keys for this Redis Cache.
+     *
+     * @param keyType key type to regenerate
+     * @return the generated access keys for this Redis Cache
+     * @throws CloudException exception thrown from REST call
+     * @throws IOException    exception thrown from serialization/deserialization
+     */
+    RedisAccessKeys regenerateKey(RedisKeyType keyType) throws CloudException, IOException;
```
Comment (anudeepsharma) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: add a blank like below (116)

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisCacheImpl.java [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+        return this.inner().redisVersion();
+    }
+
+    @Override
+    public Sku sku() {
+        return this.inner().sku();
+    }
+
+    @Override
+    public Boolean enableNonSslPort() {
```
Comment (anudeepsharma) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisCacheImpl.java]: boolean?

Comment (anudeepsharma) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisCacheImpl.java]: enableNonSslPort looks like function to enable non ssl port, is it?

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+
+    /**
+     * @return the Redis configuration value
+     */
+    Map<String, String> redisConfiguration();
+
+    /**
+     * @return the enableNonSslPort value
+     */
+    Boolean enableNonSslPort();
```
Comment (anudeepsharma) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: boolean

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisAccessKeysInner.java [hunk 5]
```diff
@@ -38,9 +38,9 @@ public String primaryKey() {
      * Set the primaryKey value.
      *
      * @param primaryKey the primaryKey value to set
-     * @return the RedisListKeysResultInner object itself.
+     * @return the RedisAccessKeysInner object itself.
      */
-    public RedisListKeysResultInner withPrimaryKey(String primaryKey) {
+    public RedisAccessKeysInner withPrimaryKey(String primaryKey) {
```
Comment (anudeepsharma) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisAccessKeysInner.java]: Most probable we are not exposing this, if yes then we should remove it as you can send in keys most probably (but I could be wrong, usually services don't give capability to set the keys you can get or regenerate).
If this is only internal, ask log a bug with swagger that these properties are read only (I am assuming we are generating withPrimary.. withSecon.. as these properties are not read only)

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisCacheImpl.java [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+            updateParameters.withSku(new Sku()
+                    .withName(SkuName.PREMIUM)
+                    .withFamily(SkuFamily.P)
+                    .withCapacity(capacity));
+        }
+        return this;
+    }
+
+    private void clearWrapperProperties() {
+        /*accountStatuses = null;
```
Comment (anudeepsharma) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisCacheImpl.java]: uncomment or remove or TODO to enable it

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisAccessKeys.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+ */
+public class RedisAccessKeys {
+    private RedisAccessKeysInner inner;
+
+    /**
+     * Creates an instance of the Redis Access keys result object.
+     *
+     * @param inner the inner object
+     */
+    public RedisAccessKeys(RedisAccessKeysInner inner) {
```
Comment (martinsawicki) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisAccessKeys.java]: why is this constructor public? Is this class intended to be instantiated by the user?  If it's only purpose is to be instantiated by our own wrappers inside the same package namespace, then it should be package-internal (no `public` keyword)

Comment (martinsawicki) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisAccessKeys.java]: why is this constructor public? Is this class intended to be instantiated by the user?  If it's only purpose is to be instantiated by our own wrappers inside the same package namespace, then it should be package-internal (no `public` keyword)

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+              * All Redis Settings. Few possible keys:
+              * rdb-backup-enabled, rdb-storage-connection-string, rdb-backup-frequency, maxmemory-delta, maxmemory-policy,
+              * notify-keyspace-events, maxmemory-samples, slowlog-log-slower-than, slowlog-max-len, list-max-ziplist-entries,
+              * list-max-ziplist-value, hash-max-ziplist-entries, hash-max-ziplist-value, set -max-intset-entries,
+              * zset-max-ziplist-entries, zset-max-ziplist-value etc.
+              *
+              * @param redisConfiguration configuration of Redis Cache as a map indexed by configuration name
+              * @return the next stage of Redis Cache definition.
+              */
+             WithCreate withRedisConfiguration(Map<String, String> redisConfiguration);
```
Comment (martinsawicki) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: if there is a specific set of strings that this method recognizes, shouldn't the key be an enum defining those possible values?  Or are arbitrary strings supported?

Comment (martinsawicki) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: if there is a specific set of strings that this method recognizes, shouldn't the key be an enum defining those possible values?  Or are arbitrary strings supported?

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+            UpdateStages.WithRedisConfiguration {
+
+        /**
+         * The full resource ID of a subnet in a virtual network to deploy the Redis Cache in.
+         * Example format: /subscriptions/{subid}/resourceGroups/{resourceGroupName}/Microsoft.{Network|ClassicNetwork}/VirtualNetworks/vnet1/subnets/subnet1
+         *
+         * @param subnetId the subnetId value to set.
+         * @return the next stage of Redis Cache update.
+         */
+        Update withSubnetId(String subnetId);
```
Comment (martinsawicki) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: we do not currently provide a way in the API to get a Subnet ID. So in cases where the code needs to point at a specific subnet reference, we let the user point at the parent network (as string resource ID ok if from a package that doesn't reference Networking), and the name of the subnet as string. For an example of that kind of reference, see `PrivateFrontend` in networking - there is `networkId()` and `subnetName()`, and then see `withExistingSubnet()` among its DefinitionStages.

Comment (martinsawicki) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: we do not currently provide a way in the API to get a Subnet ID. So in cases where the code needs to point at a specific subnet reference, we let the user point at the parent network (as string resource ID ok if from a package that doesn't reference Networking), and the name of the subnet as string. For an example of that kind of reference, see `PrivateFrontend` in networking - there is `networkId()` and `subnetName()`, and then see `withExistingSubnet()` among its DefinitionStages.

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java (hunk 2 above)
Comment (hovsepm) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: fixed

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisCacheImpl.java (hunk 3 above)
Comment (hovsepm) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisCacheImpl.java]: fixed

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java (hunk 4 above)
Comment (hovsepm) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: fixed

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisCacheImpl.java (hunk 6 above)
Comment (hovsepm) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisCacheImpl.java]: fixed

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisAccessKeysInner.java (hunk 5 above)
Comment (hovsepm) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisAccessKeysInner.java]: fixed. Changed Primary and Secondary keys to be read-only

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisAccessKeys.java (hunk 7 above)
Comment (hovsepm) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisAccessKeys.java]: fixed.

Comment (hovsepm) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisAccessKeys.java]: fixed.

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java (hunk 8 above)
Comment (hovsepm) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: please see above.

Comment (martinsawicki) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: ok.

Comment (martinsawicki) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: ok

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java (hunk 9 above)
Comment (hovsepm) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/RedisCache.java]: fixed.

File: azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisAccessKeysImpl.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+package com.microsoft.azure.management.redis.implementation;
+
+import com.microsoft.azure.management.redis.RedisAccessKeys;
+import com.microsoft.azure.management.redis.RedisCache;
+
+/**
+ * The {@link RedisCache#keys} action result.
+ */
+public class RedisAccessKeysImpl implements RedisAccessKeys {
```
Comment (martinsawicki) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisAccessKeysImpl.java]: why is this impl class public?  in general, our Impl classes are package-internal (i.e. no `public` keyword)

Comment (hovsepm) [inline azure-mgmt-redis/src/main/java/com/microsoft/azure/management/redis/implementation/RedisAccessKeysImpl.java]: fixed

**Pull request merged**